tools/funcinterval.py

```diff
@@ -0,0 +1,161 @@
+    # time the interval of do_sys_open()
+    ./funcinterval do_sys_open
+    # time the interval of xhci_ring_ep_doorbell(), in microseconds
+    ./funcinterval -u xhci_ring_ep_doorbell
+    # time the interval of do_nanosleep(), in milliseconds
+    ./funcinterval -m do_nanosleep
+    # output every 5 seconds, with timestamps
+    ./funcinterval -mTi 5 vfs_read
+    # time process 181 only
+    ./funcinterval -p 181 vfs_read
+"""
+parser = argparse.ArgumentParser(
+    description="Time interval and print latency as a histogram",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-p", "--pid", type=int,
+    help="trace this PID only")
+parser.add_argument("-i", "--interval", type=int,
+    help="summary interval, in seconds")
+parser.add_argument("-d", "--duration", type=int,
+    help="total duration of trace, in seconds")
+parser.add_argument("-T", "--timestamp", action="store_true",
+    help="include timestamp on output")
+parser.add_argument("-u", "--microseconds", action="store_true",
+    help="microsecond histogram")
+parser.add_argument("-m", "--milliseconds", action="store_true",
+    help="millisecond histogram")
+parser.add_argument("-v", "--verbose", action="store_true",
+    help="print the BPF program (for debugging purposes)")
+parser.add_argument("func_name",
+    help="Function name for tracing")
+parser.add_argument("--ebpf", action="store_true",
+    help=argparse.SUPPRESS)
+args = parser.parse_args()
+if args.duration and not args.interval:
+    args.interval = args.duration
+if not args.interval:
+    args.interval = 99999999
+
+func_name = args.func_name
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+
+BPF_HASH(start, u32, u64, 1);
+BPF_HISTOGRAM(dist);
+
+int trace_func_entry(struct pt_regs *ctx)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 index = 0, tgid = pid_tgid >> 32;
+    u64 *tsp, ts = bpf_ktime_get_ns(), delta;
+
+    FILTER
+    tsp = start.lookup(&index);
+    if (tsp == 0)
+        goto out;
+
+    delta = ts - *tsp;
+    FACTOR
+
+    // store as histogram
+    dist.increment(bpf_log2l(delta));
+
+out:
+    start.update(&index, &ts);
+
+    return 0;
+}
+"""
+
+# code substitutions
+if args.pid:
+    bpf_text = bpf_text.replace('FILTER',
+        'if (tgid != %d) { return 0; }' % args.pid)
+else:
+    bpf_text = bpf_text.replace('FILTER', '')
+if args.milliseconds:
+    bpf_text = bpf_text.replace('FACTOR', 'delta /= 1000000;')
+    label = "msecs"
+elif args.microseconds:
+    bpf_text = bpf_text.replace('FACTOR', 'delta /= 1000;')
+    label = "usecs"
+else:
+    bpf_text = bpf_text.replace('FACTOR', '')
+    label = "nsecs"
+
+if args.verbose or args.ebpf:
+    print(bpf_text)
+    if args.ebpf:
+        exit()
+
+# signal handler
+def signal_ignore(signal, frame):
+    print()
+
+
+# load BPF program
+b = BPF(text=bpf_text)
+
+# attach probes
+b.attach_kprobe(event_re=func_name, fn_name="trace_func_entry")
```

Comment:
Here, we are matching as regex. Maybe you have an option to control this? Sometimes people really just want to track a particular function.